File: apps/sidecar/DirectorSidecar/DirectorSidecar/Observability/Observability.swift
//
//  Observability.swift
//  DirectorSidecar
//

import Foundation
import OSLog

enum ObservabilityRecordKind: String, Codable, Sendable {
    case log
    case span
    case metric
    case analytics
    case error
}

enum ObservabilityLogLevel: String, Codable, Sendable {
    case debug
    case info
    case warn
    case error
}

enum ObservabilitySpanStatus: String, Codable, Sendable {
    case ok
    case error
}

enum ObservabilityAnalyticsStage: String, Codable, Sendable {
    case sessionStarted = "session_started"
    case contextSelected = "context_selected"
    case transcriptAccepted = "transcript_accepted"
    case planApproved = "plan_approved"
    case planRejected = "plan_rejected"
    case actionCompleted = "action_completed"
    case actionFailed = "action_failed"
    case interruptUsed = "interrupt_used"
}

enum ObservabilityAttributeValue: Equatable, Sendable, Codable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case null

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Double.self) {
            self = .number(value)
        } else {
            self = .string(try container.decode(String.self))
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case let .string(value): try container.encode(value)
        case let .number(value): try container.encode(value)
        case let .bool(value): try container.encode(value)
        case .null: try container.encodeNil()
        }
    }
}

enum ObservabilityPrivacyError: Error, Equatable, LocalizedError {
    case emptyField(String)
    case invalidTimestamp(String)
    case negativeDurationMs(Double)
    case nonFiniteNumber(String)
    case forbiddenAttributeKey(String)
    case missingKindField(String)

    var errorDescription: String? {
        switch self {
        case let .emptyField(field): return "observability field must not be empty: \(field)"
        case let .invalidTimestamp(timestamp): return "observability timestamp must be ISO-8601: \(timestamp)"
        case let .negativeDurationMs(value): return "observability span duration must be nonnegative: \(value)"
        case let .nonFiniteNumber(field): return "observability number must be finite: \(field)"
        case let .forbiddenAttributeKey(key): return "observability attributes must not include private/raw field name: \(key)"
        case let .missingKindField(field): return "observability record is missing required field: \(field)"
        }
    }
}

enum ObservabilityPrivacy {
    private static let forbiddenFragments = [
        "credential", "password", "prompt", "raw", "screenshot", "secret", "token",
        "transcript", "windowtitle", "windowcontent", "appname", "apptitle", "utterance",
        "screencontent",
    ]

    static func validateAttributes(_ attributes: [String: ObservabilityAttributeValue]) throws {
        for (key, value) in attributes {
            guard !key.isEmpty else { throw ObservabilityPrivacyError.emptyField("attributes.key") }
            if isForbiddenAttributeKey(key) {
                throw ObservabilityPrivacyError.forbiddenAttributeKey(key)
            }
            if case let .number(number) = value, !number.isFinite {
                throw ObservabilityPrivacyError.nonFiniteNumber("attributes.\(key)")
            }
        }
    }

    static func isForbiddenAttributeKey(_ key: String) -> Bool {
        let normalized = key
            .lowercased()
            .filter { $0.isLetter || $0.isNumber }
        return forbiddenFragments.contains { normalized.contains($0) }
    }
}

struct ObservabilityRecord: Equatable, Sendable, Codable {
    let kind: ObservabilityRecordKind
    let timestamp: String
    let component: String
    let event: String
    let release: String?
    let platform: String?
    let sessionId: String?
    let correlationId: String?
    let traceId: String?
    let spanId: String?
    let attributes: [String: ObservabilityAttributeValue]
    let level: ObservabilityLogLevel?
    let parentSpanId: String?
    let durationMs: Double?
    let status: ObservabilitySpanStatus?
    let name: String?
    let value: Double?
    let unit: String?
    let stage: ObservabilityAnalyticsStage?
    let errorClass: String?
    let handled: Bool?

    init(
        kind: ObservabilityRecordKind,
        timestamp: String,
        component: String,
        event: String,
        release: String? = nil,
        platform: String? = nil,
        sessionId: String? = nil,
        correlationId: String? = nil,
        traceId: String? = nil,
        spanId: String? = nil,
        attributes: [String: ObservabilityAttributeValue] = [:],
        level: ObservabilityLogLevel? = nil,
        parentSpanId: String? = nil,
        durationMs: Double? = nil,
        status: ObservabilitySpanStatus? = nil,
        name: String? = nil,
        value: Double? = nil,
        unit: String? = nil,
        stage: ObservabilityAnalyticsStage? = nil,
        errorClass: String? = nil,
        handled: Bool? = nil
    ) throws {
        try Self.validateBase(
            timestamp: timestamp,
            component: component,
            event: event,
            release: release,
            platform: platform,
            sessionId: sessionId,
            correlationId: correlationId,
            traceId: traceId,
            spanId: spanId
        )
        try ObservabilityPrivacy.validateAttributes(attributes)
        if let durationMs, !durationMs.isFinite { throw ObservabilityPrivacyError.nonFiniteNumber("durationMs") }
        if let durationMs, durationMs < 0 { throw ObservabilityPrivacyError.negativeDurationMs(durationMs) }
        try Self.validateKind(
            kind,
            level: level,
            status: status,
            name: name,
            value: value,
            unit: unit,
            stage: stage,
            errorClass: errorClass,
            handled: handled
        )

        self.kind = kind
        self.timestamp = timestamp
        self.component = component
        self.event = event
        self.release = release
        self.platform = platform
        self.sessionId = sessionId
        self.correlationId = correlationId
        self.traceId = traceId
        self.spanId = spanId
        self.attributes = attributes
        self.level = level
        self.parentSpanId = parentSpanId
        self.durationMs = durationMs
        self.status = status
        self.name = name
        self.value = value
        self.unit = unit
        self.stage = stage
        self.errorClass = errorClass
        self.handled = handled
    }

    private static func validateKind(
        _ kind: ObservabilityRecordKind,
        level: ObservabilityLogLevel?,
        status: ObservabilitySpanStatus?,
        name: String?,
        value: Double?,
        unit: String?,
        stage: ObservabilityAnalyticsStage?,
        errorClass: String?,
        handled: Bool?
    ) throws {
        switch kind {
        case .log:
            guard level != nil else { throw ObservabilityPrivacyError.missingKindField("level") }
        case .span:
            guard status != nil else { throw ObservabilityPrivacyError.missingKindField("status") }
        case .metric:
            try validateRequired(name, "name")
            guard let value else { throw ObservabilityPrivacyError.missingKindField("value") }
            guard value.isFinite else { throw ObservabilityPrivacyError.nonFiniteNumber("value") }
            try validateOptional(unit, "unit")
        case .analytics:
            guard stage != nil else { throw ObservabilityPrivacyError.missingKindField("stage") }
        case .error:
            try validateRequired(errorClass, "errorClass")
            guard handled != nil else { throw ObservabilityPrivacyError.missingKindField("handled") }
        }
    }

    private static func validateBase(
        timestamp: String,
        component: String,
        event: String,
        release: String?,
        platform: String?,
        sessionId: String?,
        correlationId: String?,
        traceId: String?,
        spanId: String?
    ) throws {
        guard isISODateTime(timestamp) else { throw ObservabilityPrivacyError.invalidTimestamp(timestamp) }
        try validateRequired(component, "component")
        try validateRequired(event, "event")
        try validateOptional(release, "release")
        try validateOptional(platform, "platform")
        try validateOptional(sessionId, "sessionId")
        try validateOptional(correlationId, "correlationId")
        try validateOptional(traceId, "traceId")
        try validateOptional(spanId, "spanId")
    }

    private static func validateRequired(_ value: String?, _ field: String) throws {
        guard let value, !value.isEmpty else { throw ObservabilityPrivacyError.emptyField(field) }
    }

    private static func validateOptional(_ value: String?, _ field: String) throws {
        if let value, value.isEmpty { throw ObservabilityPrivacyError.emptyField(field) }
    }

    static func isISODateTime(_ value: String) -> Bool {
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if fractional.date(from: value) != nil { return true }
        let plain = ISO8601DateFormatter()
        plain.formatOptions = [.withInternetDateTime]
        return plain.date(from: value) != nil
    }
}

enum ObservabilityExport: String, Codable, Sendable {
    case local
    case remote
}

struct ObservabilityExportPolicy: Equatable, Sendable, Codable {
    let remoteRequested: Bool
    let consentGranted: Bool

    static let localDefault = ObservabilityExportPolicy(
        remoteRequested: false,
        consentGranted: false
    )

    static func remote(consentGranted: Bool) -> ObservabilityExportPolicy {
        ObservabilityExportPolicy(remoteRequested: true, consentGranted: consentGranted)
    }

    var remoteExportAllowed: Bool {
        remoteRequested && consentGranted
    }
}

struct ObservabilityEnvelope: Equatable, Sendable, Codable {
    let schemaVersion: String
    let generatedAt: String
    let export: ObservabilityExport
    let remoteExportAllowed: Bool
    let record: ObservabilityRecord

    init(
        record: ObservabilityRecord,
        policy: ObservabilityExportPolicy = .localDefault,
        generatedAt: String
    ) throws {
        guard ObservabilityRecord.isISODateTime(generatedAt) else {
            throw ObservabilityPrivacyError.invalidTimestamp(generatedAt)
        }
        self.schemaVersion = "observability.v1"
        self.generatedAt = generatedAt
        self.export = policy.remoteExportAllowed ? .remote : .local
        self.remoteExportAllowed = policy.remoteExportAllowed
        self.record = record
    }

    static func crash(
        timestamp: String,
        component: String,
        errorClass: String,
        sessionId: String? = nil,
        attributes: [String: ObservabilityAttributeValue] = [:],
        policy: ObservabilityExportPolicy = .localDefault
    ) throws -> ObservabilityEnvelope {
        let record = try ObservabilityRecord(
            kind: .error,
            timestamp: timestamp,
            component: component,
            event: "crash",
            sessionId: sessionId,
            attributes: attributes,
            errorClass: errorClass,
            handled: false
        )
        return try ObservabilityEnvelope(record: record, policy: policy, generatedAt: timestamp)
    }
}

protocol ObservabilitySink: Sendable {
    func emit(_ envelope: ObservabilityEnvelope) async throws
}

struct ObservabilityOSLogSink: ObservabilitySink {
    private let logger: Logger

    init(
        subsystem: String = Bundle.main.bundleIdentifier ?? "DirectorSidecar",
        category: String = "observability"
    ) {
        self.logger = Logger(subsystem: subsystem, category: category)
    }

    func emit(_ envelope: ObservabilityEnvelope) async throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        let data = try encoder.encode(envelope)
        guard let line = String(data: data, encoding: .utf8) else { return }
        logger.info("\(line, privacy: .private)")
    }
}

actor ObservabilityMemorySink: ObservabilitySink {
    private let limit: Int
    private var emitted: [ObservabilityEnvelope] = []

    init(limit: Int = 256) {
        self.limit = max(1, limit)
    }

    func emit(_ envelope: ObservabilityEnvelope) async throws {
        emitted.append(envelope)
        if emitted.count > limit {
            emitted.removeFirst(emitted.count - limit)
        }
    }

    func records() -> [ObservabilityRecord] {
        emitted.map(\.record)
    }

    func envelopes() -> [ObservabilityEnvelope] {
        emitted
    }
}

struct ObservabilityClient: Sendable {
    let component: String
    let sink: any ObservabilitySink
    let policy: ObservabilityExportPolicy
    let release: String?
    let platform: String?
    let clock: @Sendable () -> String

    init(
        component: String,
        sink: any ObservabilitySink = ObservabilityMemorySink(),
        policy: ObservabilityExportPolicy = .localDefault,
        release: String? = nil,
        platform: String? = "macos",
        clock: @escaping @Sendable () -> String = { ObservabilityClient.timestamp() }
    ) {
        self.component = component
        self.sink = sink
        self.policy = policy
        self.release = release
        self.platform = platform
        self.clock = clock
    }

    func log(
        _ level: ObservabilityLogLevel,
        event: String,
        sessionId: String? = nil,
        correlationId: String? = nil,
        attributes: [String: ObservabilityAttributeValue] = [:]
    ) async throws {
        try await emit(
            kind: .log, event: event, sessionId: sessionId, correlationId: correlationId,
            attributes: attributes, level: level)
    }

    func span(
        event: String,
        sessionId: String? = nil,
        correlationId: String? = nil,
        traceId: String? = nil,
        spanId: String? = nil,
        parentSpanId: String? = nil,
        durationMs: Double? = nil,
        status: ObservabilitySpanStatus = .ok,
        attributes: [String: ObservabilityAttributeValue] = [:]
    ) async throws {
        try await emit(
            kind: .span,
            event: event,
            sessionId: sessionId,
            correlationId: correlationId,
            traceId: traceId,
            spanId: spanId,
            attributes: attributes,
            parentSpanId: parentSpanId,
            durationMs: durationMs,
            status: status
        )
    }

    func metric(
        name: String,
        value: Double,
        unit: String? = nil,
        event: String,
        sessionId: String? = nil,
        correlationId: String? = nil,
        attributes: [String: ObservabilityAttributeValue] = [:]
    ) async throws {
        try await emit(
            kind: .metric,
            event: event,
            sessionId: sessionId,
            correlationId: correlationId,
            attributes: attributes,
            name: name,
            value: value,
            unit: unit
        )
    }

    func analytics(
        stage: ObservabilityAnalyticsStage,
        event: String,
        sessionId: String? = nil,
        correlationId: String? = nil,
        attributes: [String: ObservabilityAttributeValue] = [:]
    ) async throws {
        try await emit(
            kind: .analytics, event: event, sessionId: sessionId, correlationId: correlationId,
            attributes: attributes, stage: stage)
    }

    func error(
        event: String,
        errorClass: String,
        handled: Bool,
        sessionId: String? = nil,
        correlationId: String? = nil,
        attributes: [String: ObservabilityAttributeValue] = [:]
    ) async throws {
        try await emit(
            kind: .error,
            event: event,
            sessionId: sessionId,
            correlationId: correlationId,
            attributes: attributes,
            errorClass: errorClass,
            handled: handled
        )
    }

    private func emit(
        kind: ObservabilityRecordKind,
        event: String,
        sessionId: String? = nil,
        correlationId: String? = nil,
        traceId: String? = nil,
        spanId: String? = nil,
        attributes: [String: ObservabilityAttributeValue] = [:],
        level: ObservabilityLogLevel? = nil,
        parentSpanId: String? = nil,
        durationMs: Double? = nil,
        status: ObservabilitySpanStatus? = nil,
        name: String? = nil,
        value: Double? = nil,
        unit: String? = nil,
        stage: ObservabilityAnalyticsStage? = nil,
        errorClass: String? = nil,
        handled: Bool? = nil
    ) async throws {
        let now = clock()
        let record = try ObservabilityRecord(
            kind: kind,
            timestamp: now,
            component: component,
            event: event,
            release: release,
            platform: platform,
            sessionId: sessionId,
            correlationId: correlationId,
            traceId: traceId,
            spanId: spanId,
            attributes: attributes,
            level: level,
            parentSpanId: parentSpanId,
            durationMs: durationMs,
            status: status,
            name: name,
            value: value,
            unit: unit,
            stage: stage,
            errorClass: errorClass,
            handled: handled
        )
        try await sink.emit(try ObservabilityEnvelope(record: record, policy: policy, generatedAt: now))
    }

    private static func timestamp() -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: Date())
    }

    static func bundleRelease(_ bundle: Bundle = .main) -> String? {
        let version = bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
        let build = bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String
        switch (version, build) {
        case let (.some(version), .some(build)): return "\(version)+\(build)"
        case let (.some(version), .none): return version
        case let (.none, .some(build)): return build
        case (.none, .none): return nil
        }
    }
}

File: apps/desktop/src-tauri/src/commands/observability.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use std::sync::Mutex;

const FORBIDDEN_ATTRIBUTE_KEY_PARTS: [&str; 8] = [
    "credential",
    "password",
    "prompt",
    "raw",
    "screenshot",
    "secret",
    "token",
    "transcript",
];

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ObservabilityAttributeValue {
    String(String),
    Number(f64),
    Bool(bool),
    Null,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ObservabilityBase {
    pub timestamp: String,
    pub component: String,
    pub event: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub release: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub platform: Option<String>,
    #[serde(rename = "sessionId", skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(rename = "correlationId", skip_serializing_if = "Option::is_none")]
    pub correlation_id: Option<String>,
    #[serde(rename = "traceId", skip_serializing_if = "Option::is_none")]
    pub trace_id: Option<String>,
    #[serde(rename = "spanId", skip_serializing_if = "Option::is_none")]
    pub span_id: Option<String>,
    #[serde(default)]
    pub attributes: BTreeMap<String, ObservabilityAttributeValue>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum LogLevel {
    Debug,
    Info,
    Warn,
    Error,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SpanStatus {
    Ok,
    Error,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AnalyticsStage {
    SessionStarted,
    ContextSelected,
    TranscriptAccepted,
    PlanApproved,
    PlanRejected,
    ActionCompleted,
    ActionFailed,
    InterruptUsed,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum ObservabilityRecord {
    #[serde(rename = "log", rename_all = "camelCase")]
    Log {
        #[serde(flatten)]
        base: ObservabilityBase,
        level: LogLevel,
    },
    #[serde(rename = "span", rename_all = "camelCase")]
    Span {
        #[serde(flatten)]
        base: ObservabilityBase,
        #[serde(rename = "parentSpanId", skip_serializing_if = "Option::is_none")]
        parent_span_id: Option<String>,
        #[serde(rename = "durationMs", skip_serializing_if = "Option::is_none")]
        duration_ms: Option<f64>,
        status: SpanStatus,
    },
    #[serde(rename = "metric", rename_all = "camelCase")]
    Metric {
        #[serde(flatten)]
        base: ObservabilityBase,
        name: String,
        value: f64,
        #[serde(skip_serializing_if = "Option::is_none")]
        unit: Option<String>,
    },
    #[serde(rename = "analytics", rename_all = "camelCase")]
    Analytics {
        #[serde(flatten)]
        base: ObservabilityBase,
        stage: AnalyticsStage,
    },
    #[serde(rename = "error", rename_all = "camelCase")]
    Error {
        #[serde(flatten)]
        base: ObservabilityBase,
        error_class: String,
        handled: bool,
    },
}

impl ObservabilityRecord {
    pub fn attributes_mut(&mut self) -> &mut BTreeMap<String, ObservabilityAttributeValue> {
        &mut self.base_mut().attributes
    }

    fn validate(&self) -> Result<(), String> {
        validate_base(self.base())?;
        match self {
            Self::Log { .. } | Self::Analytics { .. } => Ok(()),
            Self::Span {
                parent_span_id,
                duration_ms,
                ..
            } => {
                validate_optional_string(parent_span_id, "parentSpanId")?;
                if duration_ms.is_some_and(|value| !value.is_finite() || value < 0.0) {
                    return Err(
                        "invalid-observability-record: durationMs must be nonnegative".to_string(),
                    );
                }
                Ok(())
            }
            Self::Metric {
                name, value, unit, ..
            } => {
                validate_required_string(name, "name")?;
                validate_optional_string(unit, "unit")?;
                if !value.is_finite() {
                    return Err("invalid-observability-record: value must be finite".to_string());
                }
                Ok(())
            }
            Self::Error { error_class, .. } => {
                validate_required_string(error_class, "errorClass")?;
                Ok(())
            }
        }
    }

    fn base(&self) -> &ObservabilityBase {
        match self {
            Self::Log { base, .. }
            | Self::Span { base, .. }
            | Self::Metric { base, .. }
            | Self::Analytics { base, .. }
            | Self::Error { base, .. } => base,
        }
    }

    fn base_mut(&mut self) -> &mut ObservabilityBase {
        match self {
            Self::Log { base, .. }
            | Self::Span { base, .. }
            | Self::Metric { base, .. }
            | Self::Analytics { base, .. }
            | Self::Error { base, .. } => base,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ExportPolicy {
    LocalOnly,
    RemoteAllowed,
}

pub struct ObservabilitySink {
    export_policy: ExportPolicy,
    records: Mutex<Vec<ObservabilityRecord>>,
}

impl Default for ObservabilitySink {
    fn default() -> Self {
        Self::local_only()
    }
}

impl ObservabilitySink {
    pub fn local_only() -> Self {
        Self {
            export_policy: ExportPolicy::LocalOnly,
            records: Mutex::new(Vec::new()),
        }
    }

    pub fn with_remote_export_enabled() -> Self {
        Self {
            export_policy: ExportPolicy::RemoteAllowed,
            records: Mutex::new(Vec::new()),
        }
    }

    pub fn emit(&self, record: ObservabilityRecord) -> Result<(), String> {
        record.validate()?;
        self.records
            .lock()
            .map_err(|_| "observability sink is unavailable".to_string())?
            .push(record);
        Ok(())
    }

    pub fn records(&self) -> Vec<ObservabilityRecord> {
        self.records
            .lock()
            .expect("observability sink is unavailable")
            .clone()
    }

    pub fn export_policy(&self) -> ExportPolicy {
        self.export_policy
    }
}

fn validate_base(base: &ObservabilityBase) -> Result<(), String> {
    validate_required_string(&base.timestamp, "timestamp")?;
    validate_required_string(&base.component, "component")?;
    validate_required_string(&base.event, "event")?;
    validate_optional_string(&base.release, "release")?;
    validate_optional_string(&base.platform, "platform")?;
    validate_optional_string(&base.session_id, "sessionId")?;
    validate_optional_string(&base.correlation_id, "correlationId")?;
    validate_optional_string(&base.trace_id, "traceId")?;
    validate_optional_string(&base.span_id, "spanId")?;
    validate_attributes(&base.attributes)
}

fn validate_required_string(value: &str, field: &str) -> Result<(), String> {
    if value.trim().is_empty() {
        return Err(format!(
            "invalid-observability-record: {field} must not be empty"
        ));
    }
    Ok(())
}

fn validate_optional_string(value: &Option<String>, field: &str) -> Result<(), String> {
    if value
        .as_deref()
        .is_some_and(|value| value.trim().is_empty())
    {
        return Err(format!(
            "invalid-observability-record: {field} must not be empty"
        ));
    }
    Ok(())
}

fn validate_attributes(
    attributes: &BTreeMap<String, ObservabilityAttributeValue>,
) -> Result<(), String> {
    for (key, value) in attributes {
        validate_required_string(key, "attribute key")?;
        let normalized = key.to_ascii_lowercase();
        if FORBIDDEN_ATTRIBUTE_KEY_PARTS
            .iter()
            .any(|forbidden| normalized.contains(forbidden))
        {
            return Err(
                "invalid-observability-record: attributes must not include private/raw field names"
                    .to_string(),
            );
        }
        if matches!(value, ObservabilityAttributeValue::Number(value) if !value.is_finite()) {
            return Err(
                "invalid-observability-record: attribute numbers must be finite".to_string(),
            );
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::BTreeMap;

    const TS: &str = "2026-06-27T12:00:00Z";

    fn base(event: &str) -> ObservabilityBase {
        ObservabilityBase {
            timestamp: TS.to_string(),
            component: "desktop".to_string(),
            event: event.to_string(),
            release: None,
            platform: Some("macos".to_string()),
            session_id: Some("session-1".to_string()),
            correlation_id: Some("correlation-1".to_string()),
            trace_id: Some("trace-1".to_string()),
            span_id: Some("span-1".to_string()),
            attributes: BTreeMap::from([(
                "command".to_string(),
                ObservabilityAttributeValue::String("intent_resolve".to_string()),
            )]),
        }
    }

    #[test]
    fn emits_representative_records_to_the_memory_sink() {
        let sink = ObservabilitySink::default();
        let records = vec![
            ObservabilityRecord::Log {
                base: base("intent.request"),
                level: LogLevel::Info,
            },
            ObservabilityRecord::Span {
                base: base("intent.worker"),
                parent_span_id: Some("parent-1".to_string()),
                duration_ms: Some(42.0),
                status: SpanStatus::Ok,
            },
            ObservabilityRecord::Metric {
                base: base("intent.latency"),
                name: "intent_request_ms".to_string(),
                value: 42.0,
                unit: Some("ms".to_string()),
            },
            ObservabilityRecord::Analytics {
                base: base("plan.approved"),
                stage: AnalyticsStage::PlanApproved,
            },
            ObservabilityRecord::Error {
                base: base("intent.error"),
                error_class: "provider-unavailable".to_string(),
                handled: true,
            },
        ];

        for record in &records {
            sink.emit(record.clone()).expect("record is accepted");
        }

        assert_eq!(sink.records(), records);
    }

    #[test]
    fn serializes_contract_field_names_without_null_optionals() {
        let record = ObservabilityRecord::Span {
            base: base("intent.worker"),
            parent_span_id: Some("parent-1".to_string()),
            duration_ms: Some(42.0),
            status: SpanStatus::Ok,
        };

        assert_eq!(
            serde_json::to_value(record).expect("record serializes"),
            serde_json::json!({
                "kind": "span",
                "timestamp": TS,
                "component": "desktop",
                "event": "intent.worker",
                "platform": "macos",
                "sessionId": "session-1",
                "correlationId": "correlation-1",
                "traceId": "trace-1",
                "spanId": "span-1",
                "attributes": {
                    "command": "intent_resolve"
                },
                "parentSpanId": "parent-1",
                "durationMs": 42.0,
                "status": "ok"
            })
        );
    }

    #[test]
    fn rejects_private_or_raw_attribute_keys() {
        let sink = ObservabilitySink::default();

        for key in [
            "credential",
            "password",
            "prompt",
            "raw",
            "screenshot",
            "secret",
            "token",
            "transcript",
        ] {
            let mut record = ObservabilityRecord::Log {
                base: base("intent.request"),
                level: LogLevel::Info,
            };
            record
                .attributes_mut()
                .insert(key.to_string(), ObservabilityAttributeValue::Bool(true));

            let result = sink.emit(record);

            assert!(matches!(result, Err(message) if message.contains("private/raw")));
            assert!(sink.records().is_empty());
        }
    }

    #[test]
    fn defaults_to_local_export_disabled() {
        let sink = ObservabilitySink::default();

        assert_eq!(sink.export_policy(), ExportPolicy::LocalOnly);
    }

    #[test]
    fn remote_export_requires_explicit_opt_in() {
        let sink = ObservabilitySink::with_remote_export_enabled();

        assert_eq!(sink.export_policy(), ExportPolicy::RemoteAllowed);
    }
}
